### PR TITLE
show_bears and filter_by_language in coala-json

### PR DESCRIPTION
**Pull request by arafsheikh**

### CODE REVIEW

File: coalib/coala_json.py [hunk 1]
```diff
@@ -28,12 +33,51 @@ def main():
...
+                                                   ],
+                               "Optional settings": [param for param in
+                                                     metadata.optional_params]
+                               })
+
+            print(json.dumps({"bears": retval},
+                             cls=JSONEncoder,
+                             sort_keys=True,
+                             indent=2,
+                             separators=(',', ': ')))
```
Comment (sils1297) [inline coalib/coala_json.py]: naaah, I'd put this into an output variable or so, so it'll be printed to file if the user requests so

File: coalib/coala_json.py [hunk 2]
```diff
@@ -28,12 +33,51 @@ def main():
...
+        try:
+            sections, _ = load_configuration(arg_list=None,
+                                             log_printer=log_printer)
+            local_bears, global_bears = collect_all_bears_from_sections(
+                sections, log_printer)
+            if args.filter_by_language:
+                local_bears = filter_section_bears_by_languages(
+                    local_bears, args.filter_by_language)
+                global_bears = filter_section_bears_by_languages(
+                    global_bears, args.filter_by_language)
```
Comment (sils1297) [inline coalib/coala_json.py]: this is code duplication between coala and coala-json I think we should avoid that

Comment (sils1297) [inline coalib/coala_json.py]: see also https://github.com/coala-analyzer/coala/issues/2321 as we ight make the `--json` an argument of the main coala app, not sure how we wanna do this yet

Comment (arafsheikh) [inline coalib/coala_json.py]: I need this now for coala-eclipse.

Comment (arafsheikh) [inline coalib/coala_json.py]: How do you propose to de-duplicate?

Comment (sils1297) [inline coalib/coala_json.py]: maybe a function that does this stuff? Not fully sure, possibly ask others for ideas

**Comment by sils1297** [comment]
marking WIP in the meantimg

File: coalib/coala_json.py (hunk 2 above)
Comment (arafsheikh) [inline coalib/coala_json.py]: I'm planning to make a function out of line 40-48, but in which file should the function reside? `CliParsing.py`?

Comment (hypothesist) [inline coalib/coala_json.py]: IMO Collectors.py sounds more fitting (you're collecting bears that are for a particular language).

I'm also for making this a function as I'm doing the exact same thing in coala-quickstart unnecessarily.

File: coalib/parsing/DefaultArgParser.py [hunk 3]
```diff
@@ -187,23 +187,23 @@ def default_arg_parser(formatter_class=None):
...
-    # The following are "coala" specific arguments
-    if parser_type == 'coala':
-        SHOW_BEARS_HELP = ("Display bears and its metadata with the sections "
-                           "that they belong to")
-        outputs_group.add_argument('-B',
-                                   '--show-bears',
-                                   const=True,
-                                   action='store_const',
-                                   help=SHOW_BEARS_HELP)
+    SHOW_BEARS_HELP = ("Display bears and its metadata with the sections "
```
Comment (AbdealiJK) [inline coalib/parsing/DefaultArgParser.py]: If you remove the if condition, then this is even there for coala-format, coala-html. etc etc...
You want the if condition to check for coala and coala-json only

File: coalib/coala_json.py [hunk 4]
```diff
@@ -37,26 +35,19 @@ def main():
...
-                    global_bears, args.filter_by_language)
+            local_bears, global_bears = get_filtered_bears(
+                args.filter_by_language, log_printer)
 
             bears = inverse_dicts(local_bears, global_bears)
             results = []
-            for bear, _sections in sorted(bears.items(),
-                                          key=lambda bear_tuple:
-                                          bear_tuple[0].name):
+            for bear, sections in sorted(bears.items(),
```
Comment (sils1297) [inline coalib/coala_json.py]: put the sections rename in an own commit

File: coalib/coala_json.py [hunk 5]
```diff
@@ -37,26 +35,19 @@ def main():
...
+                                         key=lambda bear_tuple:
+                                         bear_tuple[0].name):
                 metadata = bear.get_metadata()
                 results.append({"Name": bear.name,
                                 "Description": metadata.desc,
-                                "Supported languages": bear.supported_languages,
-                                "Used in": _sections,
+                                "Supported languages": bear.LANGUAGES,
```
Comment (sils1297) [inline coalib/coala_json.py]: the switch to LANGUAGES is an own commit

File: coalib/collecting/Collectors.py [hunk 6]
```diff
@@ -228,6 +228,28 @@ def collect_all_bears_from_sections(sections, log_printer):
     return local_bears, global_bears
 
 
+def get_filtered_bears(language, log_printer):
+    """
+    Fetch bears and filter them based on given language.
```
Comment (sils1297) [inline coalib/collecting/Collectors.py]: languages please, plural. This allows a list as well.

File: coalib/collecting/Collectors.py [hunk 7]
```diff
@@ -228,6 +228,28 @@ def collect_all_bears_from_sections(sections, log_printer):
...
+
+    :param language:    Filter bears based on the given language.
+    :param log_printer: The log_printer to handle logging.
+    :return:            Tuple containing dictionaries of local bears
+                        and global bears.
+    """
+    from coalib.settings.ConfigurationGathering import load_configuration
```
Comment (sils1297) [inline coalib/collecting/Collectors.py]: why the import here? Circular import? Sounds evil. That shouldn't be!

Comment (arafsheikh) [inline coalib/collecting/Collectors.py]: Oh, was circular import the reason why I was getting an import error if I imported it with other stuff above?

Comment (sils1297) [inline coalib/collecting/Collectors.py]: guess so, that's evil can we try finding a way around it?

Comment (sils1297) [inline coalib/collecting/Collectors.py]: to me it looks like this function doesn't at all belong into this file

File: coalib/coala_json.py [hunk 8]
```diff
@@ -28,12 +31,38 @@ def main():
...
-    retval = {"results": results}
+            bears = inverse_dicts(local_bears, global_bears)
+            results = []
+            for bear, sections in sorted(bears.items(),
+                                         key=lambda bear_tuple:
+                                         bear_tuple[0].name):
+                metadata = bear.get_metadata()
+                results.append({"Name": bear.name,
+                                "Description": metadata.desc,
+                                "Supported languages": bear.LANGUAGES,
```
Comment (sils1297) [inline coalib/coala_json.py]: bears have more attributes by now, can't you just append the bear and let the JSONEncoder do the job? It should do it just fine extracting also all the other info.

Comment (arafsheikh) [inline coalib/coala_json.py]: @sils1297 If I append bear directly I get `TypeError: <property object at 0x7f71baf57318> is not JSON serializable`. Although `set([type(x) for x in dir(bear)])` says all the attributes are strings, so they should be serializable.

Comment (sils1297) [inline coalib/coala_json.py]: uhm some interactive debugging might help

File: coalib/collecting/Collectors.py (hunk 7 above)
Comment (arafsheikh) [inline coalib/collecting/Collectors.py]: There's another cyclic import [here](https://github.com/coala-analyzer/coala/blob/master/coalib/collecting/Collectors.py#L196). We can move both functions to a different module. Or maybe a new module.

Comment (sils1297) [inline coalib/collecting/Collectors.py]: move them to ConfigurationGathering would make sense wouldn't it?

Comment (arafsheikh) [inline coalib/collecting/Collectors.py]: But these functions have nothing to do with config files. I was thinking something like a `Provider`(like content/info provider) module in Collectors dir?

File: coalib/coala_json.py (hunk 8 above)
Comment (arafsheikh) [inline coalib/coala_json.py]: @AbdealiJK You wrote the `JSONEncoder`, can you help me with this?

Comment (AbdealiJK) [inline coalib/coala_json.py]: Possibly add a try/catch and catch TypeErrors. And print which property is causing this ?
Also, `bear`s have sets right ? For LANGUAGE, etc etc ... So, I don't see how everything can be a string.
`set()` is not JSON serializable:
```
In [2]: a = {"a": set([1,2])}
In [3]: a
Out[3]: {'a': {1, 2}}
In [4]: import json
In [5]: json.dumps(a)
...
TypeError: set([1, 2]) is not JSON serializable
```

Comment (arafsheikh) [inline coalib/coala_json.py]: `set` is not a problem as [this](https://github.com/coala-analyzer/coala/blob/master/coalib/output/JSONEncoder.py#L23) takes care of it.
`<class 'property'>` is the problem. More specifically https://github.com/coala-analyzer/coala/blob/master/coalib/bears/Bear.py#L386

FWIW I tired commenting out the whole function but then I got another TypeError stating `<class '_sre.SRE_Pattern'>`(which is the compiled regex object) can't be serialized

Comment (arafsheikh) [inline coalib/coala_json.py]: @sils1297 I think that appending `bear` isn't a good idea. The JSONEncoder will recursively go through all attributes in a bear, but some attributes like [this](https://github.com/coala-analyzer/coala-bears/blob/master/bears/c_languages/ClangComplexityBear.py#L23) will break the JSONEncoder. We can handle such cases in JSONEncoder but then we'll have to do this every time something isn't serializable.
I think we should stick to only outputting the stuff we want (like in `coala`). What do you say?

Comment (sils1297) [inline coalib/coala_json.py]: The then the JSONEncoder is buggy, the stuff *should* be serializable, if it is't that's a bug!

Comment (arafsheikh) [inline coalib/coala_json.py]: `CursorKind.IF_STMT, CursorKind.WHILE_STMT, ...` are external classes. We can't predict if they are serializable.

Comment (sils1297) [inline coalib/coala_json.py]: then they should be private or so and not be serialized

Comment (arafsheikh) [inline coalib/coala_json.py]: @sils1297 @AbdealiJK Okay so almost everything is fixed, except I wanted to know the `non_optional_params` for each bear for coala-eclipse. This info can be accessed using `get_metadata()` classmethod.
Is it possible to somehow append this data to each bear?

Comment (sils1297) [inline coalib/coala_json.py]: we could make the metadata a property so it gets serialized as well, sounds like something useful in general, huh?

File: coalib/bears/Bear.py [hunk 9]
```diff
@@ -261,6 +262,29 @@ def get_metadata(cls):
...
+            "desc": metadata.desc,
+            "non_optional_params": [({param: non_optional_params[param][0]}
+                                     for param in non_optional_params)],
+            "optional_params": [({param: optional_params[param][0]}
+                                 for param in optional_params)]}
+
+        # Delete attributes that cannot be serialized
+        unserializable_attributes = ["new_result", "printer"]
+        for attribute in unserializable_attributes:
+            if attribute in _dict:
```
Comment (arafsheikh) [inline coalib/bears/Bear.py]: Coverage drops as the branch will always be true in the current state. Should I remove this check and directly delete the element from the dict?

File: tests/coalaJSONTest.py [hunk 10]
```diff
@@ -51,6 +53,29 @@ def test_fail_acquire_settings(self):
...
+            output = json.loads(output)
+            self.assertEqual(len(output["bears"]), 4)
+
+    def test_show_language_bears(self):
+        with bear_test_module():
+            retval, output = execute_coala(
+                coala_json.main, 'coala-json', '-B', '-l', 'java')
+            self.assertEqual(retval, 0)
+            output = json.loads(output)
+            self.assertEqual(len(output["bears"]), 2)
```
Comment (tushar-rishav) [inline tests/coalaJSONTest.py]: @arafsheikh Could you please a test for the other attributes too? Like name, desc etc?

**Comment by tushar-rishav** [comment]
ack 460a8c190345aa29e47ac65ceb27b20a3f7e2e9b

File: tests/coalaJSONTest.py (hunk 10 above)
Comment (arafsheikh) [inline tests/coalaJSONTest.py]: @tushar-rishav Done (see below).

**Comment by tushar-rishav** [comment]
ack f9e46c8

**Comment by tushar-rishav** [comment]
unack f9e46c8

**Comment by tushar-rishav** [comment]
ack https://github.com/coala-analyzer/coala/commit/f9e46c89794035d4dad7dc2a149faf935c28a8c4

**Comment by tushar-rishav** [comment]
@rultor merge
